### PR TITLE
Add readonly connection endpoints

### LINKED ISSUE #8127: API Endpoints - Read-only - Connection
**Description**

Hello 

We need to create several endpoints that perform read-only operations on **Connection** . We need the following endpoints:

- GET /connections
- GET /connections/{connection_id}

For now, we focus only on read-only operations, but others will also be implemented as the next step.

Detailed information is available in the issue:
https://github.com/apache/airflow/issues/8118

Lots of love,

**Use case / motivation**
N/A

**Related Issues**
N/A

### PR DESCRIPTION
**Pull request by ephraimbuddy**
---
Closes: #8127 
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: airflow/api_connexion/endpoints/connection_endpoint.py [hunk 1]
```diff
@@ -47,7 +65,7 @@ def patch_connection():
     raise NotImplementedError("Not implemented yet.")
 
 
-def post_connection():
+def post_connection(session):
```
Comment (turbaszek) [inline airflow/api_connexion/endpoints/connection_endpoint.py]: Nit: this is not related to read-only endpoints

File: airflow/api_connexion/schemas/base_schema.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        d_list = []
+        for item in data:
+            for k, v in item.items():
+                if v is None and k in self.FIELDS_FROM_NONE_TO_EMPTY_STRING:
+                    v = ''
+                elif v is None and k in self.FIELDS_FROM_NONE_TO_ZERO:
+                    v = 0
+                d_data[k] = v
+            d_list.append(d_data)
+            d_data = {}
```
Comment (turbaszek) [inline airflow/api_connexion/schemas/base_schema.py]:
```suggestion
        for item in data:
            d_data = self._process_data(item)
            d_list.append(d_data)
```
The code is duplicated below so this should work. WDYT?

File: airflow/api_connexion/schemas/connection_schema.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+# under the License.
+
+from marshmallow_sqlalchemy import auto_field
+
+from airflow.api_connexion.schemas.base_schema import BaseSchema
+from airflow.models.connection import Connection
+
+
+class ConnectionCollectionItemSchema(BaseSchema):
+
```
Comment (turbaszek) [inline airflow/api_connexion/schemas/connection_schema.py]:
```suggestion
```

File: airflow/api_connexion/schemas/base_schema.py (hunk 2 above)
Comment (turbaszek) [reply airflow/api_connexion/schemas/base_schema.py]: Or even simpler:
```python
def _process_list_data(self, data):
    return [self._process_data(item) for item in data]
```

File: airflow/api_connexion/schemas/base_schema.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    def _process_data(self, data):
+        d_data = {}
+        for k, v in data.items():
+            if v is None and k in self.FIELDS_FROM_NONE_TO_EMPTY_STRING:
+                v = ''
+            elif v is None and k in self.FIELDS_FROM_NONE_TO_ZERO:
+                v = 0
+            d_data[k] = v
+        return d_data
```
Comment (turbaszek) [inline airflow/api_connexion/schemas/base_schema.py]:
```suggestion
        d_data = {}
        for k, v in data.items():
            if v is not None:
                continue
            elif k in self.FIELDS_FROM_NONE_TO_EMPTY_STRING:
                v = ''
            elif k in self.FIELDS_FROM_NONE_TO_ZERO:
                v = 0
            d_data[k] = v
        return d_data
```

WDYT?

Comment (ephraimbuddy) [reply airflow/api_connexion/schemas/base_schema.py]: I tried having it exactly like above but had issues with my tests. I have updated it similar to this

**Review by turbaszek — APPROVED** [review]
LGTM

File: airflow/api_connexion/schemas/base_schema.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        """
+        Checks if data is a list and use the envelope key to return it together
+        with total_entries meta
+        :param data: The deserialized data
+        :param many: Whether the data is a collection
+        """
+        self.check_collection_name()
+        if many:
+            data = self._process_list_data(data)
+            return {self.COLLECTION_NAME: data, 'total_entries': len(data)}
```
Comment (mik-laj) [inline airflow/api_connexion/schemas/base_schema.py]: Great idea. I wonder if pagination is handled here correctly? When we have 100 entries in the database, the user will fetch only two entries (limit = 2), then total_entires should still be 100.

File: airflow/api_connexion/schemas/base_schema.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        return data
+
+    def _process_list_data(self, data):
+        return [self._process_data(item) for item in data]
+
+    def _process_data(self, data):
+        d_data = {}
+        for k, v in data.items():
+            if v is None:
+                if k in self.FIELDS_FROM_NONE_TO_EMPTY_STRING:
```
Comment (mik-laj) [inline airflow/api_connexion/schemas/base_schema.py]: I think we should add nullable to the specification in a few places. This was not done early because we focused on the general API specification.

File: airflow/api_connexion/schemas/base_schema.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                    v = ''
+                elif k in self.FIELDS_FROM_NONE_TO_ZERO:
```
Comment (mik-laj) [inline airflow/api_connexion/schemas/base_schema.py]: It seems to me that we could do this by extending the fields.String or fields.Int class. It will be clearer.

File: tests/api_connexion/endpoints/test_connection_endpoint.py [hunk 8]
```diff
@@ -29,38 +30,99 @@ def setUpClass(cls) -> None:
...
         assert response.status_code == 200
 
 
 class TestGetConnection(TestConnectionEndpoint):
-    @pytest.mark.skip(reason="Not implemented yet")
-    def test_should_response_200(self):
-        response = self.client.get("/api/v1/connection/1")
+
+    @provide_session
+    def test_should_response_200(self, session):
```
Comment (mik-laj) [inline tests/api_connexion/endpoints/test_connection_endpoint.py:53]: Can you add support for 404 as well?

File: airflow/api_connexion/schemas/base_schema.py (hunk 7 above)
Comment (ephraimbuddy) [reply airflow/api_connexion/schemas/base_schema.py]: Good. I will try it

File: airflow/api_connexion/schemas/base_schema.py (hunk 5 above)
Comment (ephraimbuddy) [reply airflow/api_connexion/schemas/base_schema.py]: I have not tried pagination. Let me write a testcase for it

File: airflow/api_connexion/schemas/base_schema.py (hunk 6 above)
Comment (ephraimbuddy) [reply airflow/api_connexion/schemas/base_schema.py]: That would be great,

File: airflow/api_connexion/schemas/base_schema.py (hunk 5 above)
Comment (ephraimbuddy) [reply airflow/api_connexion/schemas/base_schema.py]: It works on pagination. I will push now. Please, Is there a way I can change this PR back to draft?

Comment (mik-laj) [reply airflow/api_connexion/schemas/base_schema.py]: You can convert PR to draft
Screenshot_20200604-213646

File: airflow/api_connexion/schemas/base_schema.py (hunk 7 above)
Comment (ephraimbuddy) [reply airflow/api_connexion/schemas/base_schema.py]: Hi, I was not able to fix this by extending fields but adding nullable to the specification like you said fixed it. This is the fix [819e036](https://github.com/apache/airflow/pull/9095/commits/819e036694f92e9a805941e4d464884853f24791)
Also I am thinking that the spec changes can be done in a separate PR

**Comment by vbmade2000** [comment]
A question. Should we have a directory/filename with package name in it ? Like here we have airflow/api_connexion/endpoints/connection_endpoint.py. This path contains connexion name in it. IMHO code should not reflect underlying package. Just my opinion.

**Comment by mik-laj** [comment]
@vbmade2000 Packages should clearly show the content. We already have the api package, so I had to use another package that best describes its content. api_connexion is best in my opinion, because we rely on connexion

File: airflow/api_connexion/schemas/base_schema.py (hunk 7 above)
Comment (mik-laj) [reply airflow/api_connexion/schemas/base_schema.py]: > Also I am thinking that the spec changes can be done in a separate PR

You can suggest it. For me, it's okay if the changes are proposed with context.

File: airflow/api_connexion/openapi/v1.yaml [hunk 9]
```diff
@@ -1146,14 +1146,19 @@ components:
           type: string
         conn_type:
           type: string
+          nullable: true
```
Comment (mik-laj) [inline airflow/api_connexion/openapi/v1.yaml]: This fielld should be non-nullable.  I created separate PR about the database schema.
https://github.com/apache/airflow/pull/9187

Comment (mik-laj) [reply airflow/api_connexion/openapi/v1.yaml]: Ok. Can you do rebase? My change has been merged.

Comment (ephraimbuddy) [reply airflow/api_connexion/openapi/v1.yaml]: I hurriedly pushed now. I will still come back to it this night. [e40d532](https://github.com/apache/airflow/pull/9095/commits/e40d532c01461672514fa07291633bb3796fd9b6)

File: airflow/api_connexion/openapi/v1.yaml [hunk 10]
```diff
@@ -1148,12 +1148,16 @@ components:
           type: string
         host:
           type: string
+          nullable: true
```
Comment (mik-laj) [inline airflow/api_connexion/openapi/v1.yaml:1151]: +1

**Review by mik-laj — APPROVED** [review]

**Comment by mik-laj** [comment]
Do you want to add something? I would like to merge this change, but it is a draft, so I can't do it.

**Comment by ephraimbuddy** [comment]
You can merge it. I don't have anything to add again

**Comment by turbaszek** [comment]
Thanks @ephraimbuddy !
